---
title: "Sox8/9 data"
author: "Clayton Pio Santiago"
date: "2025-06-02"
output: html_document
---
Load required packages
```{r include=FALSE}
library(Seurat)
library(Signac)
library(tidyverse)
library(viridis)
library(RColorBrewer)
library(ggpattern)
library(monocle3)
library(harmony)
library(SeuratWrappers)
library(enrichplot)
library(clusterProfiler)
library(DOSE)
library(SeuratDisk)
library(patchwork)
library(plotly)
library(scDblFinder)
library(ArchR)
library(slingshot)
library(pheatmap)
library(ComplexHeatmap)
library(cowplot)
library(scales)
library(UCell)
library(Matrix)
library(SingleCellExperiment)
library(apeglm)
library(BiocParallel)
library(nichenetr)
library(EnsDb.Mmusculus.v79)
library(BSgenome.Mmusculus.UCSC.mm10)
library(org.Mm.eg.db) 
library(TFBSTools)
library(motifmatchr)
library(JASPAR2024)
library(regioneR)
library(msigdbr)
library(fgsea)
library(rtracklayer)
library(GenomicRanges)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(EnrichedHeatmap)
library(circlize)
library(ChIPseeker)

set.seed(1234)
```

Load Nicole's multiome data
```{r include=FALSE}
sox <- readRDS(file = "F:/Sox KO/mg_multi")
```

Add motif information and motif activity
```{r include=FALSE}
#Change default assay
DefaultAssay(sox) <- "ATAC"

#Load JASPAR position frequency matrix
JASPAR2024 <- JASPAR2024()
JASPARConnect <- RSQLite::dbConnect(RSQLite::SQLite(), db(JASPAR2024))
pfm <- getMatrixSet(JASPARConnect, 
                    opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE))
RSQLite::dbDisconnect(JASPARConnect)

#Add motifs
sox <- AddMotifs(sox, genome = BSgenome.Mmusculus.UCSC.mm10, pfm = pfm)

#Run ChromVAR
sox <- RunChromVAR(sox, genome = BSgenome.Mmusculus.UCSC.mm10)

#Add TF names to ChromVAR assay meta features
sox@assays$chromvar@meta.features$motifname <- ConvertMotifID(sox, 
                                                      id = rownames(sox@assays$chromvar@meta.features))

meta_features <- sox@assays$chromvar@meta.features
meta_features <- meta_features %>%  mutate(tf1 = str_to_title(gsub("\\.\\d+$", "", sub("::.*", "", motifname))),
                                            tf2 = ifelse(grepl("::", motifname), str_to_title(gsub("\\.\\d+$", "", sub(".*::", "", motifname))), NA))
sox@assays$chromvar@meta.features <- meta_features

saveRDS(sox, file = "F:/Sox KO/mg_multi.rds")
rm (pfm, JASPAR2024, JASPARConnect, meta_features)
```

Find DE genes, DA peaks, enriched motifs and differential motif activity in Sox8/9 MGs
```{r include=FALSE}
meta_features <- sox@assays$chromvar@meta.features
meta_features$id <- rownames (meta_features)

#Find DEGs
sox_DEG <- FindMarkers(sox, ident.1 = "Sox8/9 DKO", ident.2 = "Control", 
                         assay = "RNA_temp", min.pct = 0.1, logfc.threshold = 0.2)
sox_DEG <- subset(sox_DEG, p_val_adj < 0.05)
write.csv(sox_DEG, file = "F:/Sox KO/Sox89vscon_DEG_padj05.csv")

#Find DARs
sox_DAR <- FindMarkers(sox, ident.1 = "Sox8/9 DKO", ident.2 = "Control", 
                         assay = "ATAC", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_ATAC")
annot = ClosestFeature(sox, region = rownames(sox_DAR))
sox_DAR <- sox_DAR %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)
write.csv(sox_DAR, file = "F:/Sox KO/Sox89vscon_DAR.csv")

#Find enriched motifs
sox_DAR_up <- rownames(sox_DAR[sox_DAR$p_val_adj  < 0.05 & sox_DAR$avg_log2FC > 0.2, ])
sox_DAR_down <- rownames(sox_DAR[sox_DAR$p_val_adj  < 0.05 & sox_DAR$avg_log2FC < -0.2, ])

DefaultAssay(sox) <- "ATAC"
sox_openpeaks <- AccessiblePeaks(sox, idents = c("Sox8/9 DKO"))
con_openpeaks <- AccessiblePeaks(sox, idents = c("Control"))

peakfeatures <- GetAssayData(sox, assay = "ATAC", slot = "meta.features")
sox_up_matched <- MatchRegionStats(meta.feature = peakfeatures[sox_openpeaks, ],
    query.feature = peakfeatures[sox_DAR_up, ],  n = 50000)
sox_down_matched <- MatchRegionStats(meta.feature = peakfeatures[con_openpeaks, ],
    query.feature = peakfeatures[sox_DAR_down, ],  n = 50000)

sox_enrichedmotif <- FindMotifs(sox, features = sox_DAR_up, background = sox_up_matched)
sox_enrichedmotif <- subset(sox_enrichedmotif, p.adjust < 0.005)
write.csv(sox_enrichedmotif, file = "F:/Sox KO/Sox89_enrichmotif.csv")

con_enrichedmotif <- FindMotifs(sox, features = sox_DAR_down, background = sox_down_matched)
con_enrichedmotif <- subset(con_enrichedmotif, p.adjust < 0.005)
write.csv(con_enrichedmotif, file = "F:/Sox KO/con_enrichmotif.csv")

#Find differential motif activity
sox_motifactivity <- FindMarkers(sox, ident.1 = "Sox8/9 DKO",  ident.2 = "Control",
      logfc.threshold = 0.1, min.pct=0, assay = "chromvar",
      mean.fxn = rowMeans, fc.name = "avg_diff")
sox_motifactivity$motif_name <- plyr::mapvalues(x =rownames(sox_motifactivity), 
                                       from = meta_features$id, to = meta_features$motifname)
write.csv(sox_motifactivity, file = "F:/Sox KO/Sox89vscon_diffmotifactivity.csv")
 
rm (annot, peakfeatures, sox_up_matched, sox_down_matched, 
    sox_openpeaks, con_openpeaks, sox_DAR_up, sox_DAR_down)
```

Perform GO analysis on DEGs
```{r include=FALSE}
genelist <- sox_DEG$avg_log2FC
names(genelist) <- rownames(sox_DEG)
genelist = sort(genelist, decreasing = TRUE)

#GSEA on GO terms
gse <- gseGO(geneList=genelist, 
             ont ="ALL", 
             keyType = "SYMBOL", 
             nPerm = 10000, 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Mm.eg.db, 
             pAdjustMethod = "none")

#Plot dotplot for GSEA 
dotplot(gse, showCategory=10, split=".sign") + facet_grid(.~.sign)
rm(genelist)
```

Plotting figures for paper
```{r}
#Plot UMAP
colors2 <- c("#4484DF", "#DF9144")
DimPlot(sox, reduction = "umap", cols= colors2)

#Plot volcano plots for DARs
sox_DAR_plot <- sox_DAR[sox_DAR$p_val_adj < 0.05 & (sox_DAR$avg_log2FC > 0.2 | sox_DAR$avg_log2FC < -0.2),]
sox_DAR_plot$Enriched <- "Control"
sox_DAR_plot$Enriched[sox_DAR_plot$avg_log2FC > 0.2] <- "Sox8/9 DKO"

dar_counts <- table(sox_DAR_plot$Enriched)
legend_labels <- c(
  paste0("Control (", dar_counts["Control"], " DARs)"),
  paste0("Sox8/9 DKO (", dar_counts["Sox8/9 DKO"], " DARs)")
)
names(legend_labels) <- c("Control", "Sox8/9 DKO")

ggplot(sox_DAR_plot, aes(x=avg_log2FC, y=-log10(p_val_adj), color=Enriched)) + 
  geom_point() + 
  theme_classic() +
  scale_color_manual(values=colors2, limits= c("Control", "Sox8/9 DKO"), labels = legend_labels ) +
  xlab("log2(Fold Change)") + 
  ylab("-log10(p-value)") +
  ggtitle("Differential Accessibility") +
  theme(legend.title = element_text(size = 10),
    legend.text  = element_text(size = 10),
    plot.title   = element_text(size = 12, hjust = 0.5))

rm (dar_counts, legend_labels, sox_DAR_plot)

#Plot volcano plots for DEGs
sox_DEG$Enriched <- "Control"
sox_DEG$Enriched[sox_DEG$avg_log2FC > 0.2] <- "Sox8/9 DKO"

deg_counts <- table(sox_DEG$Enriched)
legend_labels <- c(
  paste0("Control (", deg_counts["Control"], " DEGs)"),
  paste0("Sox8/9 DKO (", deg_counts["Sox8/9 DKO"], " DEGs)")
)
names(legend_labels) <- c("Control", "Sox8/9 DKO")

ggplot(sox_DEG, aes(x=avg_log2FC, y=-log10(p_val_adj), color=Enriched)) + 
  geom_point() + 
  theme_classic() +
  scale_color_manual(values=colors2, limits= c("Control", "Sox8/9 DKO"), labels = legend_labels ) +
  xlab("log2(Fold Change)") + 
  ylab("-log10(p-value)") +
  ggtitle("Differential Expressed Genes") +
  xlim(-6, 6) +
  theme(legend.title = element_text(size = 10),
    legend.text  = element_text(size = 10),
    plot.title   = element_text(size = 12, hjust = 0.5))

rm (deg_counts, legend_labels)

#Plot ViolinPlot for a few genes
genelist<- c("Glul","Kit","Kdr","Kcnj10","Cyp26a1",
             "Sox9","Sox8","Sox2","Plxna4", "Slit2", 
             "Nrg1","Robo1","Met","Gpld1", "Lef1", 
             "Wnt5b","Sox5","Sox6","Notch1","Hes5", "
             Hes1")

output_dir <- "G:/My Drive/JHU/Data/Sox_paper"

for (i in genelist) {
  p <- VlnPlot(sox, features = i, cols = colors2, assay = "RNA_temp", group.by = "orig.ident") +
       NoLegend() + labs(x = "")
  
  ggsave(filename = file.path(output_dir, paste0(i, "_violinplot.pdf")),
    plot = p, width = 6,height = 5, units = "in")
}

#Plot Coverage Plots
DefaultAssay(sox) <- "ATAC"
for (i in genelist) {
  p <- CoveragePlot(sox,region = i, extend.upstream = 2000, extend.downstream = 2000,  annotation = T, peaks = F, links = F) &  scale_fill_manual(values = colors2)
  
  ggsave(filename = file.path(output_dir, paste0(i, "_coverageplot.pdf")),
    plot = p, width = 9,height = 5, units = "in")
}

#Plot motif enrichment
sox_enrichedmotif <- sox_enrichedmotif[order(sox_enrichedmotif$p.adjust),]
sox_enrichedmotif$rank <- seq_len(nrow(sox_enrichedmotif))
sox_enrichedmotif$mlog10padj <- -log10(sox_enrichedmotif$p.adjust)

ggplot(sox_enrichedmotif, aes(rank, mlog10padj, color = mlog10padj)) + 
  geom_point(size = 1) +
  ggrepel::geom_label_repel(
        data = sox_enrichedmotif[rev(seq_len(15)), ], aes(x = rank, y = mlog10padj, label = motif.name), 
        size = 1.5,
        nudge_x = 2,
        color = "black",
        max.overlaps = Inf) + 
  theme_classic() + 
  ylab("-log10(p-value)") + 
  xlab("Rank Sorted Enriched Motifs") +
  scale_color_gradientn(colors = paletteContinuous(set = "comet")) + 
  ggtitle("Sox8/9 DKO Enriched Motifs") + 
  NoLegend()


con_enrichedmotif <- con_enrichedmotif[order(con_enrichedmotif$p.adjust),]
con_enrichedmotif$rank <- seq_len(nrow(con_enrichedmotif))
con_enrichedmotif$mlog10padj <- -log10(con_enrichedmotif$p.adjust)

ggplot(con_enrichedmotif, aes(rank, mlog10padj, color = mlog10padj)) + 
  geom_point(size = 1) +
  ggrepel::geom_label_repel(
        data = con_enrichedmotif[rev(seq_len(15)), ], aes(x = rank, y = mlog10padj, label = motif.name), 
        size = 1.5,
        nudge_x = 2,
        color = "black",
        max.overlaps = Inf) + 
  theme_classic() + 
  ylab("-log10(p-value)") + 
  xlab("Rank Sorted Enriched Motifs") +
  scale_color_gradientn(colors = paletteContinuous(set = "comet")) + 
  ggtitle("Control Enriched Motifs") + 
  NoLegend()

#Plot chromvar heatmap
DefaultAssay(sox) <- "chromvar"
sox_motifactivity_filt <- subset(sox_motifactivity, p_val_adj < 0.05 & abs(avg_diff) > 0.2)
write.csv(sox_motifactivity_filt, file = "F:/Sox KO/Sox89vscon_diffmotifactivity_02diff.csv")
chromvar_mat <- GetAssayData(sox, slot = "data")[rownames(sox_motifactivity_filt), ]
group_labels <- sox$orig.ident
group_levels <- c("Control", "Sox8/9 DKO") 

avg_activity <- sapply(group_levels, function(grp) {
  rowMeans(chromvar_mat[, group_labels == grp, drop = FALSE])
})
colnames(avg_activity) <- group_levels

motif_order <- sox_motifactivity_filt %>%
  rownames_to_column(var = "motif") %>%     # Save rownames to a column
  arrange(desc(avg_diff)) %>%
  pull(motif)

avg_activity <- avg_activity[motif_order, ]

pheatmap(
    avg_activity,
    cluster_rows = FALSE,
    cluster_cols = FALSE,
    show_rownames = T,
    main = "Differential Motif Activity",
    fontsize_row = 6,
    color = colorRampPalette(c("#67BFBC", "white", "#F19B26"))(100))

```

```{r}
DefaultAssay(sox) <- "ATAC"

motiflist <- c("MA0868.3", "MA0077.2", "MA0726.2", "MA0700.3", 
               "MA0718.2","MA1150.2", "MA0488.2", "MA0742.2")
sox <- Footprint(sox,
  motif.name = motiflist,
  genome = BSgenome.Mmusculus.UCSC.mm10, assay = "ATAC", in.peaks = TRUE, verbose = TRUE)

PlotFootprint(sox, features = motiflist) &  scale_fill_manual(values = colors2)
```

```{r}
#Save RNA counts as h5ad object
DefaultAssay(sox) <- "RNA_temp"

rna <- CreateSeuratObject(counts = sox@assays$RNA_temp@counts, meta.data = sox@meta.data)
rna[["RNA"]] <- as(object = rna[["RNA"]], Class = "Assay")
SaveH5Seurat(rna, filename = "C:/Users/Clayton/Documents/Globus_download/sox89_rnaassay.h5Seurat", overwrite = T, verbose = T)
Convert("C:/Users/Clayton/Documents/Globus_download/sox89_rnaassay.h5Seurat", dest = "h5ad", overwrite = T)

#Extract metadata and counts
meta <- sox@meta.data
atac_counts <- sox@assays$ATAC@counts

#Save barcodes only as a new column
meta$barcodes <- sub(".*?_", "", rownames(meta))

#Format atac_counts rownames to resemble UCSC annotations
rownames(atac_counts) <- gsub("^(.*?)-", "\\1:", rownames(atac_counts))

writeMM(atac_counts, file = "C:/Users/Clayton/Documents/Globus_download/sox89_atac_counts.mtx")
write.table(row.names(atac_counts), file = "C:/Users/Clayton/Documents/Globus_download/sox89_atac_counts_rownames.txt", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(colnames(atac_counts), file = "C:/Users/Clayton/Documents/Globus_download/sox89_atac_counts_colnames.txt", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.csv(sox@reductions$umap@cell.embeddings,
          file = "C:/Users/Clayton/Documents/Globus_download/sox89_atac_umap.csv")
write.csv(meta, file = "C:/Users/Clayton/Documents/Globus_download/sox89_atac_meta.csv")
```

Analyze P14 and P120 Chx10Cre Sox9 scRNA-seq data
```{r include=FALSE}
#Load Sox9 count data
p14_soxdata <- Read10X("F:/Sox KO/Raw/scR89/filtered_feature_bc_matrix")
sox9ko_p14 <- CreateSeuratObject(counts = p14_soxdata, project = "chx10sox9_p14", min.cells = 3, min.features = 200)
sox9ko_p14$age <- "P14"

p120_soxdata <- Read10X("F:/Sox KO/Raw/scR94/filtered_feature_bc_matrix")
sox9ko_p120 <- CreateSeuratObject(counts = p120_soxdata, project = "chx10sox9_p120", min.cells = 3, min.features = 200)
sox9ko_p120$age <- "P120"

#Merge
sox9ko_merge <-merge(sox9ko_p14,y = c(sox9ko_p120), add.cell.ids = c("sox9ko_p14","sox9ko_p120"))
sox9ko_merge$genotype <- "Chx10Cre_Sox9"

rm (p14_soxdata,p120_soxdata,sox9ko_p14,sox9ko_p120)
sox9ko_merge$percent.mt <- PercentageFeatureSet(object = sox9ko_merge, pattern = "^mt")
sox9ko_merge <- subset(sox9ko_merge, subset = percent.mt < 10)

#Identify doublets
sox9ko_merge[["RNA"]] <- JoinLayers(sox9ko_merge[["RNA"]]) #if using Seurat v5
sox9ko_merge <- NormalizeData(sox9ko_merge)

sce <- as.SingleCellExperiment(sox9ko_merge , assay = "RNA") #remove assay if using Seurat v4
sce <- scDblFinder(sce, samples="orig.ident",BPPARAM=SerialParam(RNGseed = 1234)) #compatible with windows machines
sox9ko_merge@meta.data$doublet <- plyr::mapvalues(x = rownames(sox9ko_merge@meta.data), 
                                           from = rownames(colData(sce)), to = as.character(sce$scDblFinder.class))
sox9ko_merge@meta.data$doublet <- factor(sox9ko_merge@meta.data$doublet, levels = c("singlet","doublet"))
sox9ko_merge@meta.data$doublet_score <- plyr::mapvalues(x = rownames(sox9ko_merge@meta.data), 
                                                 from = rownames(colData(sce)), to = sce$scDblFinder.score)
sox9ko_merge@meta.data$doublet_score<-as.numeric(sox9ko_merge@meta.data$doublet_score)
rm (sce)

#Subset poor quality cells
sox9ko_merge <- subset(sox9ko_merge, subset = nCount_RNA < 50000 & nCount_RNA > 1000 & doublet == "singlet")

#Run default pipeline
sox9ko_merge <- NormalizeData(sox9ko_merge) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(vars.to.regress = c("nCount_RNA","nFeature_RNA")) %>% RunPCA(verbose = FALSE)
sox9ko_merge <- RunHarmony(sox9ko_merge, group.by.vars = c("orig.ident"))
sox9ko_merge <- RunUMAP(sox9ko_merge, reduction = "harmony", dims = 1:10)
sox9ko_merge <- FindNeighbors(sox9ko_merge, reduction = "harmony", dims = 1:10)
sox9ko_merge <- FindClusters(sox9ko_merge, resolution = 1.2)

#Identify cluster
clus<-c("0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15",
        "16","17","18","19","20","21","22","23","24","25","26","27","28","29","30")
cell<-c("Rod","Rod","Rod","Rod","Rod","AC","BC","BC","AC","MG","BC","AC","Cone",
        "BC","Rod","MG","BC","AC","Doublet","AC","AC","BC","AC","V/E Cells","Microglia",
        "AC","Doublet","AC","Cone","BC","AC")
sox9ko_merge@meta.data$Celltype <- plyr::mapvalues(x = sox9ko_merge@meta.data$seurat_clusters, 
                                             from = clus, to = cell)
Idents(sox9ko_merge) <- "Celltype"

rgc <- CellSelector (FeaturePlot(sox9ko_merge, features = "Rbpms", order = T))
hc <- CellSelector (FeaturePlot(sox9ko_merge, features = "Lhx1", order = T))
Idents (sox9ko_merge, cells = rgc) <- "RGC"
Idents (sox9ko_merge, cells = hc) <- "HC"
sox9ko_merge[["Celltype"]] <- Idents(sox9ko_merge)
sox9ko_merge <- subset (sox9ko_merge, idents = "Doublet", invert = T)

order<-c("Rod","Cone","BC","AC","HC","RGC","MG","Microglia","V/E Cells")
sox9ko_merge@meta.data$Celltype <- factor(sox9ko_merge@meta.data$Celltype, levels = order)
Idents(sox9ko_merge) <- "Celltype"

rm (rgc, hc, cell, clus, order)

saveRDS(sox9ko_merge, file = "F:/Sox KO/chx10cresox9.rds")

# Get P14 and P120 control data from Li 2024 Mouse Retinal Atlas
# Create new seurat object since original object only has Ensembl IDs
li2024 <- readRDS(file = "F:/Mouse Retinal Atlas/li2024.rds")
Idents(li2024) <- "age"
li2024 <- subset (li2024, idents = c("P14","P120"))

count_matrix <- li2024@assays$RNA@counts
rownames(count_matrix) <- li2024@assays$RNA@meta.features$gene_symbols
rna <- CreateAssayObject(counts = count_matrix)
li2024[["RNA"]] <- rna
rm (gene_symbols, rna, count_matrix)

li2024 <- NormalizeData(li2024) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(vars.to.regress = c("nCount_RNA","nFeature_RNA")) %>% RunPCA(verbose = FALSE)
li2024 <- RunHarmony(li2024, group.by.vars = c("sampleid"))
li2024 <- RunUMAP(li2024, reduction = "harmony", dims = 1:10)
Idents (li2024) <- "majorclass"
li2024 <- RenameIdents(li2024, 'Endothelial' = 'V/E Cells', 'Pericyte' = 'V/E Cells')
li2024[["Celltype"]] <- Idents(li2024)
colnames(li2024@meta.data)[colnames(li2024@meta.data) == "sampleid"] <- "orig.ident"
li2024$genotype <- "Wildtype"

#Merge Sox9 and Li 2024 data
wt_sox <-merge(sox9ko_merge,y = li2024)

wt_sox <- NormalizeData(wt_sox) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(vars.to.regress = c("nCount_RNA","nFeature_RNA")) %>% RunPCA(verbose = FALSE)
wt_sox <- RunHarmony(wt_sox, group.by.vars = c("orig.ident","genotype"))
wt_sox <- RunUMAP(wt_sox, reduction = "harmony", dims = 1:10)

saveRDS(wt_sox, file = "F:/Sox KO/li2024_chx10cresox9_int.rds")

VlnPlot(wt_sox, features = c(c("Glul","Sox2","Sox5","Sox9","Lhx2","Apoe","Kcnj10","Ascl1","Ccnd1","Hes1","Gpld1","Wnt5b","Met","Slit2","Plxna4","Nrg1")), group.by = "genotype", idents = "MG", ncol = 4)
```

CUT&RUN Analysis
Create consensus Sox8 peak set
```{r}
#Read bed files from seacr output
bed_files <- list.files(path = "E:/Sox KO/Cutandrun/bed/scear_outs", 
                        pattern = "*.bed", full.names = TRUE)
granges_list <- lapply(bed_files, function(f) {
  df <- read.table(f, header = FALSE, sep = "\t", stringsAsFactors = FALSE)
  gr <- GRanges(seqnames = df[[1]],
    ranges   = IRanges(start = df[[2]], end = df[[3]]))
  if (ncol(df) > 3) {
    mcols(gr) <- df[, -(1:3), drop = FALSE]
  }
  gr
})
names(granges_list) <- gsub("\\.bed$", "", basename(bed_files))

#Get overlapping bed files for Sox8 and plot venn diagrams
gr1 <- granges_list[["p2_sox8_ab3.seacr.consensus.peak_counts"]]
gr1_n <- length(gr1)
gr2 <- granges_list[["p2_sox8_ab5.seacr.consensus.peak_counts"]]
gr2_n <- length(gr2)
gr3 <- granges_list[["p17_sox8_ab3.seacr.consensus.peak_counts"]]
gr3_n <- length(gr3)
gr4 <- granges_list[["p17_sox8_ab5.seacr.consensus.peak_counts"]]
gr4_n <- length(gr4)

hits1 <- findOverlaps(gr1, gr2)
hits2 <- findOverlaps(gr3, gr4)
overlap_n1 <- length(unique(queryHits(hits1))) 
overlap_n2 <- length(unique(queryHits(hits2)))

venn_counts1 <- c("Sox8_ab3" = gr1_n - overlap_n1,
                 "Sox8_ab5" = gr2_n - overlap_n1,
                 "Overlap" = overlap_n1)
venn_counts2 <- c("Sox8_ab3" = gr3_n - overlap_n2,
                 "Sox8_ab5" = gr4_n - overlap_n2,
                 "Overlap" = overlap_n2)
grid.newpage()
draw.pairwise.venn(
  area1 = gr1_n,
  area2 = gr2_n,
  cross.area = overlap_n1,
  category = c("Sox8_ab3", "Sox8_ab5"),
  fill = c("skyblue", "salmon"),
  alpha = 0.5,
  cex = 2,
  cat.cex = 1.5,
  scaled = FALSE,
  cat.pos = c(-45, 45), 
  cat.dist = c(0.05, 0.05))
grid.text("Overlap of Sox8 Peaks (P2)", y = unit(0.95, "npc"), 
          gp = gpar(fontsize = 16, fontface = "bold"))

grid.newpage()
draw.pairwise.venn(
  area1 = gr3_n,
  area2 = gr4_n,
  cross.area = overlap_n2,
  category = c("Sox8_ab3", "Sox8_ab5"),
  fill = c("skyblue", "salmon"),
  alpha = 0.5,
  cex = 2,
  cat.cex = 1.5,
  scaled = FALSE,
  cat.pos = c(-45, 45), 
  cat.dist = c(0.05, 0.05))
grid.text("Overlap of Sox8 Peaks (P17)", y = unit(0.95, "npc"), 
          gp = gpar(fontsize = 16, fontface = "bold"))

#Save new Sox8 overlap CUT&RUN bed files
overlap_gr1 <- pintersect(gr1[queryHits(hits1)], gr2[subjectHits(hits1)])
export(overlap_gr1, "E:/Sox KO/Cutandrun/bed/p2_sox8.bed")
overlap_gr2 <- pintersect(gr3[queryHits(hits2)], gr4[subjectHits(hits2)])
export(overlap_gr2, "E:/Sox KO/Cutandrun/bed/p17_sox8.bed")
rm (gr1_n,gr2_n,gr3_n,gr4_n,gr1,gr2,gr3,gr4,overlap_n1,overlap_n2,venn_counts1,venn_counts2,
    hits1,hits2,overlap_gr1,overlap_gr2)
```

Find enriched motifs for each peak set
```{r}
bed_files <- list.files(path = "E:/Sox KO/Cutandrun/bed/", pattern = "*.bed", full.names = TRUE)
peak_names <- gsub(".bed", "", basename(bed_files))
granges_list <- lapply(bed_files, rtracklayer::import)
names(granges_list) <- peak_names

#Load JASPAR PFMs
JASPAR2024 <- JASPAR2024()
JASPARConnect <- RSQLite::dbConnect(RSQLite::SQLite(), db(JASPAR2024))
pfm <- getMatrixSet(JASPARConnect, 
                    opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE))
RSQLite::dbDisconnect(JASPARConnect)
#sox_pfms <- pfm[grepl("SOX", name(pfm), ignore.case = TRUE)]

#Find enriched motifs in CUT&Run data
get_background <- function(gr) {randomizeRegions(gr, genome = "mm10")}
test_motif_enrichment <- function(gr, motifs, genome=BSgenome.Mmusculus.UCSC.mm10) {
  bg <- get_background(gr)
  motif_fg <- matchMotifs(motifs, gr, genome = genome)
  motif_bg <- matchMotifs(motifs, bg, genome = genome)
  fg_counts <- colSums(motifMatches(motif_fg))
  bg_counts <- colSums(motifMatches(motif_bg))
  res <- data.frame(
    motif = names(fg_counts),
    fg = fg_counts,
    bg = bg_counts,
    fg_total = length(gr),
    bg_total = length(bg))
  res$pval <- apply(res, 1, function(x) {
    fisher.test(matrix(c(
      as.numeric(x["fg"]), as.numeric(x["fg_total"]) - as.numeric(x["fg"]),
      as.numeric(x["bg"]), as.numeric(x["bg_total"]) - as.numeric(x["bg"])
    ), nrow = 2))$p.value
  })
  res$padj <- p.adjust(res$pval, method = "BH")
  res
}
enrich_results <- lapply(names(granges_list), function(nm) {
  gr <- granges_list[[nm]]
  res <- test_motif_enrichment(gr, pfm)
  res$bed_file <- nm
  res
})

enrich_df <- do.call(rbind, enrich_results)
enrich_df<-subset(enrich_df, padj < 0.05)

enrich_df <- enrich_df %>%
  left_join(meta_features[, c("id", "motifname")], by = c("motif" = "id")) %>%
  rename(motif_name = motifname)


rm (p17_soxmg_DEG,p17_soxmg_DAR,dev,sox_multi,JASPAR2024,JASPARConnect,pfm,
    get_background,test_motif_enrichment,enrich_results,bed_files,peak_names)
```

Plot Enrichment heatmaps
```{r}
promoters <- promoters(genes(TxDb.Mmusculus.UCSC.mm10.knownGene),
                 upstream = 1000, downstream = 1000)
#Load and average bigwigs
bw_dir <- "E:/Sox KO/Cutandrun/bigwig"
bw_groups <- list(
   P2_IgG = c("p2_igg_ctrl_R1.bigWig", "p2_igg_ctrl_R2.bigWig"),
   P17_IgG = c("p17_igg_ctrl_R1.bigWig"),
   P2_H3K4me3 = c("p2_h3k4me3_R1.bigWig", "p2_h3k4me3_R2.bigWig"),
   P17_H3K4me3 = c("p17_h3k4me3_R1.bigWig", "p17_h3k4me3_R1.bigWig"),
   P2_H3K27ac = c("p2_h3k27ac_R1.bigWig", "p2_h3k27ac_R2.bigWig"),
   P17_H3K27ac = c("p17_h3k27ac_R1.bigWig", "p17_h3k27ac_R2.bigWig"),
   P2_Sox8 = c("p2_sox8_ab3_R1.bigWig", "p2_sox8_ab3_R2.bigWig","p2_sox8_ab5_R1.bigWig","p2_sox8_ab5_R2.bigWig"),
   P17_Sox8 = c("p17_sox8_ab3_R1.bigWig", "p17_sox8_ab5_R1.bigWig"),
   P2_Sox9 = c("p2_sox9_ab4_R1.bigWig", "p2_sox9_ab4_R2.bigWig"),
   P17_Sox9 = c("p17_sox9_ab4_R1.bigWig"))

avg_bw <- lapply(bw_groups, function(files) {
  bw_paths <- file.path(bw_dir, files)
  bw_list <- lapply(bw_paths, import, as = "RleList")
  Reduce("+", bw_list) / length(bw_list)})

#Normalize to promoters
mat_list <- lapply(avg_bw, function(bw) {
  normalizeToMatrix(bw, target = promoters, value_column = "score", 
          mean_mode = "w0", w = 2000, extend = 1000, smooth = TRUE)})

#Scale
mat_list <- lapply(mat_list, function(m) {
  m_scaled <- t(scale(t(m)))  
  m_scaled[is.na(m_scaled)] <- 0
  m_scaled})

#Plot heatmap
col_fun <- colorRamp2(c(-2, 0, 2), c("#063267", "white", "#6D0223"))

ht_list <- NULL
for (i in seq_along(mat_list)) {
  ht_list <- ht_list +
    EnrichedHeatmap(
      mat_list[[i]],
      name = names(mat_list)[i],
      column_title = paste0(names(mat_list)[i], " (±1kb TSS)"),
      col = col_fun,
      use_raster = TRUE)}

ddraw(ht_list,
  main_heatmap = names(mat_list)[1],
  column_title = "TSS-centered enrichment (±1kb)",
  heatmap_legend_side = "right",
  annotation_legend_side = "right")

rm(promoters,bw_dir,bw_groups,avg_bw,mat_list,col_fun)
```

#Find overlap of Sox8 and Sox9 peaks between P2 and P17
```{r}
sets_to_plot <- granges_list[c("p2_sox8","p2_sox9","p17_sox8","p17_sox9")]
all_peaks <- reduce(unlist(GRangesList(sets_to_plot)))
presence_matrix <- sapply(sets_to_plot, function(gr) {
  overlaps <- findOverlaps(all_peaks, gr, type = "any")
  as.integer(1:length(all_peaks) %in% queryHits(overlaps))
})
presence_df <- as.data.frame(presence_matrix)
keep_combos <- list(
  "p2_sox8"             = c(1,0,0,0),
  "p2_sox9"             = c(0,1,0,0),
  "p17_sox8"            = c(0,0,1,0),
  "p17_sox9"            = c(0,0,0,1),
  "p2_sox8&p2_sox9"     = c(1,1,0,0),
  "p17_sox8&p17_sox9"   = c(0,0,1,1),
  "p2_sox8&p17_sox8"    = c(1,0,1,0),
  "p2_sox9&p17_sox9"    = c(0,1,0,1),
  "all_four"            = c(1,1,1,1)
)

row_patterns <- apply(presence_df, 1, paste, collapse = "")
allowed_patterns <- sapply(keep_combos, paste, collapse = "")
keep_idx <- row_patterns %in% allowed_patterns
filtered_df <- presence_df[keep_idx, ]

upset(filtered_df,
      sets = c("p2_sox8","p2_sox9","p17_sox8","p17_sox9"),
      nsets = 4,
      keep.order = TRUE,
      order.by = "freq",
      main.bar.color = "gray23",
      sets.bar.color = "gray23",
      text.scale = 1.5)

#Merge Sox8 and Sox9 as single bed files
p2_merged <- reduce(c(granges_list[["p2_sox8"]], granges_list[["p2_sox9"]]))
p17_merged <- reduce(c(granges_list[["p17_sox8"]], granges_list[["p17_sox9"]]))

# Export as BED
export(p2_merged, "E:/Sox KO/Cutandrun/bed/p2_sox8_sox9.bed")
export(p17_merged, "E:/Sox KO/Cutandrun/bed/p17_sox8_sox9.bed")

rm (sets_to_plot,all_peaks,presence_matrix,presence_df,
    row_patterns,allowed_patterns,keep_idx,filtered_df,
    keep_combos)
```

Compare P2 and P17 Sox8/9 CUT&RUN to dev ATAC and Sox8/9 cKO data
```{r}
#Get DEGs and DARs from P2 RPCs and P17 MGs for Control and Sox8/9 cKO
dev <- readRDS(file = "E:/NFI/R files/2272024/mmdev_atac_rnaimpute_counts.rds")
dev$age_Celltype <- paste(dev$age, dev$Celltype, sep="_")
Idents (dev) <- "age_Celltype"
p2_rpc_DAR <- FindMarkers(dev, ident.1 = c("P2_Late RPC","P2_nRPC"),
                          assay = "Peaks", min.pct = 0.1)
p2_rpc_DAR <- subset(p2_rpc_DAR, subset = p_val_adj < 0.05 & avg_log2FC > 0)

sox_multi <- readRDS(file = "E:/Sox KO/mg_multi.rds")
p17_soxmg_DAR <- FindMarkers(sox_multi, ident.1 = c("Sox8/9 DKO"), 
                             assay = "ATAC", min.pct = 0.1) 
p17_soxmg_DAR <- subset(p17_soxmg_DAR, subset = p_val_adj < 0.05)
p17_conup_DAR <- subset(p17_soxmg_DAR, subset = avg_log2FC < 0)
p17_soxup_DAR <- subset(p17_soxmg_DAR, subset = avg_log2FC > 0)

p17_soxmg_DEG <- FindMarkers(sox_multi, ident.1 = c("Sox8/9 DKO"), 
                             assay = "RNA_temp", min.pct = 0.1, logfc.threshold = 0.25) 
p17_soxmg_DEG <- subset(p17_soxmg_DEG, subset = p_val_adj < 0.05)
p17_conup_DEG <- subset(p17_soxmg_DEG, subset = avg_log2FC < 0)
p17_soxup_DEG <- subset(p17_soxmg_DEG, subset = avg_log2FC > 0)




```

Extract barcodes to extract celltype specific reads
In R
```{r}
#Define the cell types and samples of interest
sample_names <- c("chx10sox9_p14", "chx10sox9_p120")
cell_type <- "MG"
output_dir <- "F:/Sox KO/mg_barcodes/"

#Filter metadata for selected cell type
metadata <- sox9ko_merge@meta.data %>% filter(Celltype == cell_type)

# Initialize list to store barcodes
barcodes_by_sample <- list()

#Loop through each sample and extract barcodes
for (sample in sample_names) {
  sample_barcodes <- metadata %>%
    filter(grepl(paste0("^", sample), orig.ident)) %>%
    rownames() %>%
    gsub(".*_", "", .)
  
  barcodes_by_sample[[sample]] <- sample_barcodes

}

#Save barcodes to file
for (sample in names(barcodes_by_sample)) {
  writeLines(
    barcodes_by_sample[[sample]],
    file.path(output_dir, paste0(sample, "_", cell_type, ".txt"))
  )
}

rm(list = c("sample_names", "cell_type", "output_dir", "metadata", "barcodes_by_sample", "sample", "sample_barcodes"))

```

In linux
```{bash}
#Convert from windows to unix formatting
for file in *.txt; do
    sed -i 's/\r$//' "$file"
done

#Subset MG reads (an example)
samtools view -H ./outs/possorted_genome_bam.bam > header.sam
samtools view ./outs/possorted_genome_bam.bam   | grep -F -f <(awk '{print "CB:Z:" $1}' chx10sox9_p120_MG.txt) > body.sam
cat header.sam body.sam | samtools view -b -o sox9_p120_mg_filtered.bam
samtools index sox9_p120_mg_filtered.bam
```
